### ISSUE TITLE
Properties excluded through spread + destructuring assignment should be allowed to be unused

### ISSUE THREAD

**Issue by ide** [babel/babel-eslint#95]
Sample code is the easiest way for me to explain this. Say we have an object with a bunch of fields and you want to clone the object but exclude a couple of properties. We'd traditionally write:

``` js
var clone = Object.assign({}, originalObject);
delete clone.field1;
delete clone.field2;
```

With ES7 we can write:

``` js
var {field1, field2, ...clone} = originalObject;
```

Babel itself handles this correctly, but the linter complains if `field1` and `field2` are unused. Since the variable declarations have a meaningful side effect, I believe some code in the pipeline needs to tell eslint to ignore the unused variable declaration with `/*eslint-disable no-unused-vars */`, or perhaps Babel should omit the variable declaration if it lexically detects that the variable is used (although this may be undesirable if it breaks eval).

**Comment by sebmck** [comment]
You could probably patch escope (similar to #75) to register a binding as referenced. I've stemmed away from touching escope because of how hacky it'd be but I'm open to accepting PRs.

**Comment by AlexKVal** [comment]
And how about legitimately unused vars ?
What if the user has just removed another line of code ?
```js
var {unusedVar, usedVar, ...clone} = originalObject;
// someFunction(usedVar); // removed
```
How could linter know which var is legitimately unused and which is just forgotten?

**Comment by sebmck** [comment]
@AlexKVal Left side properties of the object rest *should* be assumed to be "used". This is also noted in the actual [proposal](https://github.com/sebmarkbage/ecmascript-rest-spread/blob/master/Issues.md#linters-complain-about-unused-variables) which offers an alternate suggestion of just using `_` and configuring your linter like so.

**Comment by ide** [comment]
@AlexKVal that code is still correct. Without the comment, two things were happening: `usedVar` was excluded from `clone` and `usedVar` was referenced to pass it to the function. With the comment, `usedVar` is still excluded from `clone`.

So in that sense, `usedVar` is actually still being used (that is, it's used to exclude the property from `clone`) even though it is no longer referenced. Does that distinction help clear things up?

**Comment by AlexKVal** [comment]
Yeah. Thanks guys !  :cherries: 
Your answers about this particular case helps not only me.

**Comment by danielberndt** [comment]
I'd actually prefer the stricter version from the proposal @sebmck talked about.
i.e. `var {unusedVar1: _, usedVar, unusedVar2: _, ...clone} = originalObject;`

this way the intent of the variables is very explicit and will cause less errors (I assume)

**Comment by sebmck** [comment]
@danielberndt Then you have duplicate declarations.

**Comment by danielberndt** [comment]
I feel it's acceptable to suppress the duplicate warning if 

1) the variable name is `_`
2) and the declaration happens within a destructuring context with a rest spread

that's at least my interpretation of @sebmarkbage's proposal

**Comment by sebmarkbage** [comment]
This is only ok when you use `var` since `let` and `const` would not allow you to use duplicate declarations.

**Comment by hzoo** [comment]
Was going to start this, but after I updated ESLint (0.22.1) and thus escope (3.1.0) with babel-eslint (3.1.10) I don't get the `no-unused-vars` error for field1 and field2 anymore? Not sure exactly what changed fixed it.

**Comment by hzoo** [comment]
Is it still the case for you @ide?

**Comment by ide** [comment]
@hzoo it's fixed for me too (eslint 0.22.1, babel-eslint 3.1.11). Thanks for the notice.

**Comment by hzoo** [comment]
Ok the fix for https://github.com/babel/babel-eslint/issues/120 made this not work again (I forgot to add a test for this).

**Comment by jedwards1211** [comment]
I would be equally happy if it were possible to mark variables used via comment:
```es6
const {field1, field2, field3, ...clone} = originalObject
// eslint-ignore-unused field2, field3
const x = field1 + ...
```
This is longer but even more explicit and less questionable than
```es6
import _ from 'lodash'
const {field1, field2: _, field3: _, ...clone} = originalObject
_.forEach(clone, ...) // uh...
```

**Comment by idolize** [comment]
If we could get the old behavior back that would be ideal IMO

**Comment by danielberndt** [comment]
Here's a solution that you can use today and that works around the issue quite well in my opinion:

you need to add this eslint-rule

```
    "no-unused-vars": [1, {
      "vars": "local",
      "varsIgnorePattern": "^_",
      "args": "after-used",
      "argsIgnorePattern": "^_"
    }]
```

which allows you to write code like this without warning
```
const {theme: _, children: _1, onSubmit: _2, ...rest} = this.props
```

**Comment by rojobuffalo** [comment]
@danielberndt I was able to get the desired result with only:
```
    no-unused-vars: [1, {
        varsIgnorePattern: "^_",
    }]
```

are those other settings necessary for just this use case?

**Comment by jedwards1211** [comment]
These days I just do
```js
const {
  field1,
  // eslint-disable no-unused-vars
  field2, field3,
  // eslint-enable no-unused-vars
  ...rest
} = props
```

It may be ugly, but I still like it better than using `varsIgnorePattern`.

**Comment by danez** [comment]
This is now fixed via https://github.com/eslint/eslint/pull/7968

**Comment by hzoo** [comment]
YES this was the worst 😛
